Data/ByteString/Internal.hs

```diff
@@ -460,25 +460,61 @@ append (PS fp1 off1 len1) (PS fp2 off2 len2) =
       withForeignPtr fp1 $ \p1 -> memcpy destptr1 (p1 `plusPtr` off1) len1
       withForeignPtr fp2 $ \p2 -> memcpy destptr2 (p2 `plusPtr` off2) len2
 
+-- | Concatenate a set of 'ByteString's.
+--
+-- The idea here is we first do a pass over the input list to determine,
+--   1. is a copy necessary? e.g. @concat []@, @concat [mempty, "hello"]@,
+--      and @concat ["hello", mempty, mempty]@ can all be handled without
+--      copying.
+--   2. if a copy is necessary, how large is the result going to be?
+--
+-- If a copy is necessary then we create a buffer of the appropriate size
+-- and do another pass over the input list, copying the chunks into the
+-- buffer. Also, since foreign calls aren't entirely free we skip over empty
+-- chunks while copying.
+--
 concat :: [ByteString] -> ByteString
-concat []     = mempty
-concat [bs]   = bs
-concat bss0   = unsafeCreate totalLen $ \ptr -> go bss0 ptr
+concat bss0 = goLen0 bss0
   where
-    totalLen = checkedSum "concat" [ len | (PS _ _ len) <- bss0 ]
-    go []                  !_   = return ()
-    go (PS fp off len:bss) !ptr = do
+    -- It's still possible that the result is empty
+    goLen0 []                     = mempty
+    goLen0 (PS _ _ 0     :bss)    = goLen0 bss
+    goLen0 (bs           :bss)    = goLen1 bs bss
+
+    -- It's still possible that the result is a single chunk
+    goLen1 bs []                  = bs
+    goLen1 bs (PS _ _ 0  :bss)    = goLen1 bs bss
+    goLen1 bs (PS _ _ len:bss)    = goLen (checkedAdd "concat" len' len) bss
+      where PS _ _ len' = bs
+
+    -- General case, just find the total length we'll need
+    goLen !total (PS _ _ len:bss) = goLen total' bss
+      where total' = checkedAdd "concat" total len
+    goLen total [] =
+      unsafeCreate total $ \ptr -> goCopy bss0 ptr
```

Comment:
In principle we could do better here I think. This use of `bss0` here will cause all of the `go` functions to be allocated as closures, since they capture `bss0`.